File: packages/ai-engine/src/strategies/tree-of-thought.ts
/**
 * TreeOfThoughtStrategy — Explores multiple reasoning branches for ambiguous tasks
 *
 * Instead of following one path, generates multiple "thought branches",
 * evaluates each, and selects the most promising one. Useful when the
 * task has multiple valid approaches and it's unclear which is best.
 *
 * @author Subash Karki
 */
import type {
  ActivationScore,
  ReasoningStrategy,
  StrategyInput,
  StrategyOutput,
  TaskContext,
} from '../types/strategy.js';
import { applyPriorFailurePenalty } from './prior-penalty.js';

// ---------------------------------------------------------------------------
// Branch types
// ---------------------------------------------------------------------------

interface BranchScore {
  feasibility: number;
  risk: number;
  effort: number;
  combined: number;
}

interface ThoughtBranch {
  id: number;
  approach: string;
  prompt: string;
  score: BranchScore;
}

// ---------------------------------------------------------------------------
// Strategy
// ---------------------------------------------------------------------------

export class TreeOfThoughtStrategy implements ReasoningStrategy {
  readonly id = 'tree-of-thought';
  readonly name = 'Tree of Thoughts';
  readonly version = '1.0.0';
  readonly description =
    'Explores multiple reasoning branches for ambiguous tasks, evaluates each, and selects the most promising one.';

  private readonly numBranches = 3;

  shouldActivate(context: TaskContext): ActivationScore {
    const { complexity, risk, isAmbiguous } = context;

    let base: ActivationScore;

    if (isAmbiguous && (complexity === 'moderate' || complexity === 'complex' || complexity === 'critical')) {
      base = {
        score: 0.85,
        reason: `Ambiguous task with ${complexity} complexity — tree-of-thought exploration recommended`,
      };
    } else if (isAmbiguous && complexity === 'simple') {
      base = {
        score: 0.5,
        reason: 'Ambiguous but simple task — tree-of-thought may help but is not critical',
      };
    } else if (
      (complexity === 'complex' || complexity === 'critical') &&
      (risk === 'medium' || risk === 'high' || risk === 'critical')
    ) {
      base = {
        score: 0.6,
        reason: `Complex task (${complexity}) with ${risk} risk — multiple approaches worth exploring`,
      };
    } else {
      base = {
        score: 0.1,
        reason: 'Simple or low-risk task — single path sufficient',
      };
    }

    return applyPriorFailurePenalty(base, this.id, context);
  }

  async execute(input: StrategyInput): Promise<StrategyOutput> {
    const start = Date.now();
    const { context, graphContext } = input;

    // 1. Generate thought branches
    const branches = this.generateBranches(context, graphContext);

    // 2. Score each branch
    const scoredBranches = branches.map((branch) => ({
      ...branch,
      score: this.scoreBranch(branch, context),
    }));

    // 3. Select the best branch
    let selectedIndex = 0;
    let highestCombined = -1;
    for (let i = 0; i < scoredBranches.length; i++) {
      if (scoredBranches[i].score.combined > highestCombined) {
        highestCombined = scoredBranches[i].score.combined;
        selectedIndex = i;
      }
    }

    const selected = scoredBranches[selectedIndex];

    // 4. Format the result
    const result = JSON.stringify({
      type: 'tree-of-thought',
      goal: context.goal,
      selectedBranch: {
        id: selected.id,
        approach: selected.approach,
        score: selected.score,
      },
      alternatives: scoredBranches
        .filter((_, i) => i !== selectedIndex)
        .map((b) => ({
          id: b.id,
          approach: b.approach,
          score: b.score,
        })),
    });

    // Confidence is the average of the winning branch's scores
    const confidence =
      (selected.score.feasibility + (1 - selected.score.risk) + (1 - selected.score.effort)) / 3;

    return {
      strategyId: this.id,
      result,
      confidence: Math.round(confidence * 100) / 100,
      tokensUsed: 0, // No LLM call — placeholder
      durationMs: Date.now() - start,
      artifacts: {
        strategy: 'tree-of-thought',
        branches: scoredBranches.map((b) => ({
          id: b.id,
          approach: b.approach,
          score: b.score,
        })),
        selectedBranch: selected.id,
        branchScores: scoredBranches.map((b) => b.score),
      },
    };
  }

  // ---------------------------------------------------------------------------
  // Internal helpers
  // ---------------------------------------------------------------------------

  private generateBranches(
    context: TaskContext,
    graphContext: StrategyInput['graphContext'],
  ): ThoughtBranch[] {
    const filePaths = graphContext.files.map((f) => f.path).join(', ');

    const approaches = [
      'Direct approach — implement as described',
      'Conservative approach — minimal changes, maximize reuse',
      'Refactoring approach — restructure then implement',
    ];

    return approaches.slice(0, this.numBranches).map((approach, i) => ({
      id: i + 1,
      approach,
      prompt: [
        `Branch ${i + 1}: ${approach}`,
        `Context files: ${filePaths || 'none'}`,
        `Goal: ${context.goal}`,
        'Evaluate: feasibility (0-1), risk (0-1), effort (0-1)',
      ].join('\n'),
      score: { feasibility: 0, risk: 0, effort: 0, combined: 0 },
    }));
  }

  private scoreBranch(branch: ThoughtBranch, context: TaskContext): BranchScore {
    const riskMultiplier =
      context.risk === 'critical'
        ? 0.8
        : context.risk === 'high'
          ? 0.6
          : context.risk === 'medium'
            ? 0.4
            : 0.2;

    const blastRadiusRisk = Math.min(1, context.blastRadius / 20);

    switch (branch.id) {
      case 1: {
        // Direct approach
        const feasibility = 0.8;
        const risk = Math.min(1, riskMultiplier + blastRadiusRisk * 0.3);
        const effort = 0.5;
        const combined = (feasibility + (1 - risk) + (1 - effort)) / 3;
        return { feasibility, risk: Math.round(risk * 100) / 100, effort, combined: Math.round(combined * 100) / 100 };
      }
      case 2: {
        // Conservative approach
        const feasibility = 0.9;
        const risk = 0.2;
        const effort = 0.3;
        const combined = (feasibility + (1 - risk) + (1 - effort)) / 3;
        return { feasibility, risk, effort, combined: Math.round(combined * 100) / 100 };
      }
      case 3: {
        // Refactoring approach
        const feasibility = 0.6;
        const risk = 0.5;
        const effort = 0.8;
        const combined = (feasibility + (1 - risk) + (1 - effort)) / 3;
        return { feasibility, risk, effort, combined: Math.round(combined * 100) / 100 };
      }
      default: {
        return { feasibility: 0.5, risk: 0.5, effort: 0.5, combined: 0.5 };
      }
    }
  }
}

File: v2/internal/ai/strategies/tree_of_thought.go
// Package strategies provides AI prompt enhancement strategies.
// TreeOfThoughtStrategy explores multiple candidate approaches, scores each
// deterministically, and surfaces the highest-scoring branch alongside the
// alternatives. v1 generated the alternatives via an LLM and re-scored them
// via a second LLM call; v2 generates 3 fixed approaches (direct, conservative,
// refactor) and scores them with the same heuristic v1 used as a fallback.
//
// Author: Subash Karki
package strategies

import (
	"fmt"
	"math"
	"strings"
)

// TreeOfThoughtStrategy explores multiple branches for ambiguous tasks.
type TreeOfThoughtStrategy struct{}

// NewTreeOfThoughtStrategy creates a TreeOfThoughtStrategy.
func NewTreeOfThoughtStrategy() *TreeOfThoughtStrategy { return &TreeOfThoughtStrategy{} }

// ID returns the unique identifier for this strategy.
func (s *TreeOfThoughtStrategy) ID() string { return "tree-of-thought" }

// Name returns a human-readable name.
func (s *TreeOfThoughtStrategy) Name() string { return "Tree of Thoughts" }

// Description returns a brief explanation of what this strategy does.
func (s *TreeOfThoughtStrategy) Description() string {
	return "Explores multiple reasoning branches for ambiguous tasks, evaluates each, and selects the most promising one."
}

// ShouldActivate scores how strongly this strategy fits the task.
// Mirrors v1: ambiguous + (moderate|complex|critical) → 0.85,
// ambiguous + simple → 0.5, complex+risky → 0.6.
func (s *TreeOfThoughtStrategy) ShouldActivate(t TaskAssessment) (float64, string) {
	switch {
	case t.IsAmbiguous && (t.Complexity == Moderate || t.Complexity == Complex || t.Complexity == Critical):
		return 0.85, fmt.Sprintf("Ambiguous task with %s complexity — tree-of-thought exploration recommended", t.Complexity)
	case t.IsAmbiguous && t.Complexity == Simple:
		return 0.5, "Ambiguous but simple task — tree-of-thought may help but is not critical"
	case (t.Complexity == Complex || t.Complexity == Critical) &&
		(t.Risk == MediumRisk || t.Risk == HighRisk || t.Risk == CriticalRisk):
		return 0.6, fmt.Sprintf("Complex task (%s) with %s risk — multiple approaches worth exploring", t.Complexity, t.Risk)
	default:
		return 0.1, "Simple or low-risk task — single path sufficient"
	}
}

// Enrich emits a structured tree-of-thought prompt with three approaches and
// the deterministic winner pre-computed.
func (s *TreeOfThoughtStrategy) Enrich(message string, t TaskAssessment, graphCtx string) string {
	branches := scoreBranches(t)
	winner := branches[0]

	var lines []string
	for _, b := range branches {
		marker := "  "
		if b.id == winner.id {
			marker = "* "
		}
		lines = append(lines, fmt.Sprintf("%sBranch %d (%s): feasibility=%.2f risk=%.2f effort=%.2f combined=%.2f",
			marker, b.id, b.approach, b.feasibility, b.risk, b.effort, b.combined))
	}

	guidance := fmt.Sprintf(`<strategy-guidance kind="tree-of-thought">
Task profile: complexity=%s, risk=%s, ambiguity=%.1f, blastRadius=%d
Approaches considered (* = recommended):
%s
Recommendation:
1. Start with the marked branch ("%s"). Re-evaluate the score if assumptions change.
2. Keep the alternatives in mind — fall back if blockers appear.
%s</strategy-guidance>

%s`, t.Complexity, t.Risk, t.AmbiguityScore, t.BlastRadius,
		strings.Join(lines, "\n"), winner.approach, graphCtx, message)
	return guidance
}

// thoughtBranch holds a candidate approach and its scores.
type thoughtBranch struct {
	id          int
	approach    string
	feasibility float64
	risk        float64
	effort      float64
	combined    float64
}

// scoreBranches generates the same 3 fixed approaches v1 used and scores them
// using the deterministic fallback heuristic. Returned highest-combined first.
func scoreBranches(t TaskAssessment) []thoughtBranch {
	riskMult := 0.2
	switch t.Risk {
	case CriticalRisk:
		riskMult = 0.8
	case HighRisk:
		riskMult = 0.6
	case MediumRisk:
		riskMult = 0.4
	}
	blastRisk := math.Min(1.0, float64(t.BlastRadius)/20.0)

	branches := []thoughtBranch{
		{
			id:          1,
			approach:    "Direct — implement as described",
			feasibility: 0.8,
			risk:        round2(math.Min(1.0, riskMult+blastRisk*0.3)),
			effort:      0.5,
		},
		{
			id:          2,
			approach:    "Conservative — minimal changes, maximize reuse",
			feasibility: 0.9,
			risk:        0.2,
			effort:      0.3,
		},
		{
			id:          3,
			approach:    "Refactoring — restructure, then implement",
			feasibility: 0.6,
			risk:        0.5,
			effort:      0.8,
		},
	}
	for i := range branches {
		b := &branches[i]
		b.combined = round2((b.feasibility + (1 - b.risk) + (1 - b.effort)) / 3)
	}
	// Sort: highest combined first.
	for i := 0; i < len(branches); i++ {
		for j := i + 1; j < len(branches); j++ {
			if branches[j].combined > branches[i].combined {
				branches[i], branches[j] = branches[j], branches[i]
			}
		}
	}
	return branches
}

// round2 rounds a float to 2 decimal places (matches v1 Math.round(x*100)/100).
func round2(v float64) float64 {
	return math.Round(v*100) / 100
}
